drivers/i2c/i2c_ll_stm32_device.h.jinja2

```diff
@@ -0,0 +1,23 @@
+{% import 'device_declare.jinja2' as i2c_stm32 %}
+
+{% set config_struct_body = """
+        .i2c = (I2C_TypeDef *){{ '%x'|format(data.device.get_reg_addr()) }},
+        .pclken = {
+                .enr = {{ data.device.get_property('clocks/0/bits') }},
+                .bus = {{ data.device.get_property('clocks/0/bus') }},
+        },
+#ifdef CONFIG_I2C_STM32_INTERRUPT
+        .irq_config_func = {{ data.gen_variables.irq_config_function_name }},
+#endif
+        .bitrate = {{ data.device.get_property('clock-frequency') }},""" %} 
+
+{% set params = {'irq_flag'           : 'CONFIG_I2C_STM32_INTERRUPT',
+                 'irq_func'           : 'stm32_i2c_isr',
+                 'config_struct'      : 'i2c_stm32_config',
+                 'data_struct'        : 'i2c_stm32_data',
+                 'api_struct'         : 'api_funcs',
+                 'init_function'      : 'i2c_stm32_init',
+		 'config_struct_body' :  config_struct_body} %}
+
+{{ i2c_stm32.device(data, ['st,stm32-i2c-v1', 'st,stm32-i2c-v2'], params) }}
```

Comment:
One question on particular multi compatible support.

As you could see this file is compatible with 2 STM32 I2C IP versions. Particularly, both versions use the same code for devices.
Then, what if they used slightly different code for devices V1 and devices V2.
For instance: V1 and V2 uses same data_struct_body but have different config_struct_body

How would you support that if I still want to have device instantiation all in once place? (Assumption is only one compatible would be used for a specific board, either V1 or V2, but for all types of boards, I want to use this same file.)
This is doable with codegen since we provide 'compatible' in API, so we only have to define data_struct_body,   config_struct_body_v1 and config_struct_body_v2, then calling API twice:
-first  with devicedeclare('compat_v1', 'config_struct_body_v1', 'data_struct_body'), 
-then with devicedeclare('compat_v2', 'config_struct_body_v2', 'data_struct_body')

Reply:
> This is doable with codegen since we provide 'compatible' in API, so we only have to define data_struct_body, 
> config_struct_body_v1 and config_struct_body_v2, then calling API twice:
> -first with devicedeclare('compat_v1', 'config_struct_body_v1', 'data_struct_body'),
> -then with devicedeclare('compat_v2', 'config_struct_body_v2', 'data_struct_body')

That is not entirely true, there is a lot more parameters.

```
devicedeclare.device_declare( \
 *     ['st,stm32-exti', ],
 *     'CONFIG_KERNEL_INIT_PRIORITY_DEVICE',
 *     'PRE_KERNEL_1',
 *     None,
 *     'stm32_exti_init',
 *     None,
 *     device_data,
 *     None)
```
So what I did in my jinja2 solution is to move the parameters out of the function call and into an object, I know it's personal taste, but I think it is cleaner. But your use-case is totally supported, it's just a matter of creating another parameter object, that points to the other config_struct_body for the second device, and then do two calls as in the codegen solution.

Reply:
> That is not entirely true, there is a lot more parameters.

Indeed, my goal was not to oversimplify codegen solution. I should have used following syntax:
devicedeclare('compat_v1', 'config_struct_body_v1', 'data_struct_body', ....)

> But your use-case is totally supported

Ok, great removing my NACK. But tbh I'd prefer to have things available (even as a prototype) rather than doable and will be done if RFC is approved (even if I do trust you).